spec/unit/run.spec.js

```diff
@@ -17,22 +17,217 @@
+            emulatorSpyObj.list_started.and.returnValue(Q.resolve([]));
+            emulatorSpyObj.list_images.and.returnValue(Q.resolve(emulatorList));
+            emulatorSpyObj.start.and.returnValue(Q.resolve(emulatorList[2].id));
+
+            return run.run().then(() => {
+                expect(emulatorSpyObj.start).toHaveBeenCalledWith(emulatorList[2].name);
+                expect(emulatorSpyObj.resolveTarget).toHaveBeenCalledWith(emulatorList[2].id);
+            });
+        });
+
+        it('should throw an error if target is specified but does not exist', done => {
+            const emulatorList = [{ name: 'emu1', id: 1 }];
+            const deviceList = ['device1'];
+            const target = 'nonexistentdevice';
+            getInstallTargetSpy.and.returnValue(target);
+
+            deviceSpyObj.list.and.returnValue(Q.resolve(deviceList));
+            emulatorSpyObj.list_started.and.returnValue(Q.resolve([]));
+            emulatorSpyObj.list_images.and.returnValue(Q.resolve(emulatorList));
+
+            run.run().then(() => {
+                done.fail('Expected error to be thrown');
+            })
+            .fail(err => {
+                expect(err).toContain(target);
+                done();
+            });
+        });
+
+        it('should build the app if a target has been found', () => {
+            getInstallTargetSpy.and.returnValue('--device');
+            deviceSpyObj.resolveTarget.and.returnValue({ target: 'device1', isEmulator: false });
+
+            return run.run().then(() => {
+                expect(buildSpyObj.run).toHaveBeenCalled();
+            });
+        });
+
+        it('should install on device after build', () => {
+            const deviceTarget = { target: 'device1', isEmulator: false };
+            getInstallTargetSpy.and.returnValue('--device');
+            deviceSpyObj.resolveTarget.and.returnValue(deviceTarget);
+
+            return run.run().then(() => {
+                expect(deviceSpyObj.install).toHaveBeenCalledWith(deviceTarget, undefined);
+            });
+        });
+
+        it('should install on emulator after build', () => {
+            const emulatorTarget = { target: 'emu1', isEmulator: true };
+            getInstallTargetSpy.and.returnValue('--emulator');
+            emulatorSpyObj.list_started.and.returnValue(Q.resolve([emulatorTarget.target]));
+            emulatorSpyObj.resolveTarget.and.returnValue(emulatorTarget);
+            emulatorSpyObj.wait_for_boot.and.returnValue(Q.resolve());
+
+            return run.run().then(() => {
+                expect(emulatorSpyObj.install).toHaveBeenCalledWith(emulatorTarget, undefined);
+            });
+        });
+    });
+
+    describe('help', () => {
+        it('should print out usage and help', () => {
+            let message = '';
+            spyOn(console, 'log').and.callFake(msg => { message += msg; });
+            spyOn(process, 'exit');
```

Comment:
This clobbers the global `console.log` and `process.exit` for the whole process. I'm not too comfortable with that. Could cause very obscure errors for other tests.

Reply:
I was worried about that when writing this, and looked into how to unspy a function in Jasmine, but it seems that Jasmine automatically unspies after a test is run. So, `console.log` and `process.exit` would not be spied in other functions. Personally I would prefer to explicitly unspy to avoid this confusion, but from what I can find, there is no way to do that.

Reply:
Interesting, I did not know that spies get unspied. OTOH it's a good thing, but OTOH I agree that explicit operations are preferable at times. Conveys intent and informs unsuspecting users like me :sweat_smile: